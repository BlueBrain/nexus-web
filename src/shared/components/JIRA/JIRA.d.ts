declare const AuthorizeJiraUI: ({
  jiraAuthUrl,
  onSubmitVerificationCode: connect,
}: {
  jiraAuthUrl: string;
  onSubmitVerificationCode: (verificationCode: string) => void;
}) => JSX.Element;
export declare const CreateIssueUI: ({
  displayType,
  projects,
  onOk,
  onCancel,
}: {
  projects: any[];
  onOk: (project: string, summary: string, description: string) => void;
  onCancel: () => void;
  displayType: 'project' | 'resource';
}) => JSX.Element;
export declare const LinkIssueUI: ({
  onOk,
  onCancel,
  searchJiraLink,
}: {
  onOk: (issueUrl: string) => void;
  onCancel: () => void;
  searchJiraLink: string;
}) => JSX.Element;
declare type JIRAPluginUIProps = {
  projects: any[];
  issues: any[];
  onCreateIssue: (
    project: string,
    summary: string,
    description: string
  ) => void;
  onLinkIssue: (issueKey: string) => void;
  onUnlinkIssue: (issueKey: string) => void;
  searchJiraLink: string;
  displayType: 'resource' | 'project';
  onNavigateToResource?: (resourceSelfUrl: string) => void;
  isLoading: boolean;
};
declare const JIRAPluginUI: ({
  projects,
  issues,
  onCreateIssue,
  onLinkIssue,
  onUnlinkIssue,
  searchJiraLink,
  displayType,
  onNavigateToResource,
  isLoading,
}: JIRAPluginUIProps) => JSX.Element;
export default JIRAPluginUI;
export { AuthorizeJiraUI };
