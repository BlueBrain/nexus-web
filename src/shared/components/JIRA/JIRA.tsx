import { DownOutlined } from '@ant-design/icons';
import {
  Button,
  Dropdown,
  Empty,
  Form,
  Input,
  Menu,
  Modal,
  Select,
  Spin,
  Table,
} from 'antd';
import TextArea from 'antd/lib/input/TextArea';
import * as React from 'react';
import { getFriendlyTimeAgoString } from '../../utils';

const AuthorizeJiraUI = ({
  jiraAuthUrl,
  onSubmitVerificationCode: connect,
}: {
  jiraAuthUrl: string;
  onSubmitVerificationCode: (verificationCode: string) => void;
}) => {
  const [verificationCode, setVerificationCode] = React.useState('');
  return (
    <>
      <h1>We first need to connect your Jira account.</h1>

      <ol>
        <li>
          Visit Jira to log-in and authorize access to Jira from Nexus Fusion
          <br />
          <br />
          <a href={jiraAuthUrl} target="_blank">
            Authorize Jira access
          </a>
          <br />
          <br />
        </li>
        <li>
          Copy the <em>verification code</em> from step 1 into the text box
          below and click Connect.
          <br />
          <br />
          <div style={{ display: 'flex' }}>
            <div>
              <Input
                type="text"
                placeholder="Verification code"
                style={{ backgroundColor: '#fff', width: 200 }}
                value={verificationCode}
                onChange={e => setVerificationCode(e.currentTarget.value)}
              />
            </div>
            <div>
              <Button type="primary" onClick={() => connect(verificationCode)}>
                Connect
              </Button>
            </div>
          </div>
        </li>
      </ol>
    </>
  );
};

export const CreateIssueUI = ({
  displayType,
  projects,
  onOk,
  onCancel,
}: {
  projects: any[];
  onOk: (project: string, summary: string, description: string) => void;
  onCancel: () => void;
  displayType: 'project' | 'resource';
}) => {
  const [summary, setSummary] = React.useState('');
  const [project, setProject] = React.useState('');
  const [description, setDescription] = React.useState('');

  const formItemLayout = {
    labelCol: {
      xs: { span: 24 },
      sm: { span: 5 },
    },
    wrapperCol: {
      xs: { span: 24 },
      sm: { span: 19 },
    },
  };
  const formItemLayoutWithOutLabel = {
    wrapperCol: {
      xs: { span: 24, offset: 0 },
      sm: { span: 19, offset: 5 },
    },
  };

  return (
    <Modal
      footer={null}
      title="Create A JIRA Issue"
      open={true}
      forceRender={true}
      destroyOnClose={true}
      mask={false}
      onCancel={() => onCancel()}
    >
      <p>
        A Jira issue will be created and linked to this Nexus{' '}
        {displayType === 'project' ? 'project' : 'resource'}
      </p>
      <Form onFinish={() => onOk(project, summary, description)}>
        <Form.Item
          label="Project"
          name="project"
          rules={[
            {
              required: true,
            },
          ]}
          {...formItemLayout}
        >
          <Select
            showSearch
            style={{ width: '100%' }}
            onChange={(value: string) => setProject(value)}
            filterOption={(input, option) =>
              // @ts-ignore
              option?.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
            }
            data-testid="project-select"
          >
            {projects.map(project => (
              <Select.Option key={project.key} value={project.key}>
                {project.name}
              </Select.Option>
            ))}
          </Select>
        </Form.Item>
        <Form.Item
          label="Summary"
          name="summary"
          rules={[
            {
              required: true,
            },
          ]}
          {...formItemLayout}
        >
          <Input
            type="text"
            value={summary}
            onChange={e => setSummary(e.currentTarget.value)}
            placeholder="Issue Summary"
          />
        </Form.Item>
        <Form.Item>
          <Form.Item label="Description" name="description" {...formItemLayout}>
            <TextArea
              value={description}
              onChange={e => setDescription(e.currentTarget.value)}
              placeholder="Issue description"
              autoSize={{ minRows: 3, maxRows: 5 }}
            />
          </Form.Item>
        </Form.Item>
        <Form.Item {...formItemLayoutWithOutLabel}>
          <Button danger onClick={() => onCancel()}>
            Cancel
          </Button>
          <Button
            type="primary"
            htmlType="submit"
            style={{ marginLeft: '20px' }}
            name="submitIssue"
          >
            Create
          </Button>
        </Form.Item>
      </Form>
    </Modal>
  );
};

export const LinkIssueUI = ({
  onOk,
  onCancel,
  searchJiraLink,
}: {
  onOk: (issueUrl: string) => void;
  onCancel: () => void;
  searchJiraLink: string;
}) => {
  const [issueKey, setIssueUrl] = React.useState('');
  return (
    <>
      <Modal
        title="Link Issue"
        open={true}
        forceRender={true}
        destroyOnClose={true}
        mask={false}
        onOk={() => onOk(issueKey)}
        onCancel={() => onCancel()}
      >
        <a href={searchJiraLink} target="_blank">
          Search for issue in Jira
        </a>
        <br />
        <br />
        <Input
          type="text"
          value={issueKey}
          onChange={e => setIssueUrl(e.currentTarget.value)}
          placeholder="Jira Issue URL"
        />
      </Modal>
    </>
  );
};

type JIRAPluginUIProps = {
  projects: any[];
  issues: any[];
  onCreateIssue: (
    project: string,
    summary: string,
    description: string
  ) => void;
  onLinkIssue: (issueKey: string) => void;
  onUnlinkIssue: (issueKey: string) => void;
  searchJiraLink: string;
  displayType: 'resource' | 'project';
  onNavigateToResource?: (resourceSelfUrl: string) => void;
  isLoading: boolean;
};
const JIRAPluginUI = ({
  projects,
  issues,
  onCreateIssue,
  onLinkIssue,
  onUnlinkIssue,
  searchJiraLink,
  displayType,
  onNavigateToResource,
  isLoading,
}: JIRAPluginUIProps) => {
  const [createIssueVisible, setCreateIssueVisible] = React.useState(false);
  const [linkIssueVisible, setLinkIssueVisible] = React.useState(false);

  const confirmUnlinkIssue = (issueKey: string) => {
    Modal.confirm({
      title: 'Unlink Issue',
      content:
        'Are you sure you want to unlink the issue from this Nexus resource?',
      onOk: () => {
        onUnlinkIssue(issueKey);
        return Promise.resolve();
      },
    });
  };

  if (isLoading) {
    return (
      <div
        style={{
          width: '100%',
          display: 'flex',
          marginBottom: '10px',
          marginTop: '10px',
          height: '200px',
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <Spin tip="Loading..." spinning={true} size="large"></Spin>
      </div>
    );
  }

  return (
    <>
      {createIssueVisible && (
        <CreateIssueUI
          displayType={displayType}
          projects={projects}
          onOk={(project, summary, description) => {
            onCreateIssue(project, summary, description);
            setCreateIssueVisible(false);
          }}
          onCancel={() => setCreateIssueVisible(false)}
        />
      )}
      {linkIssueVisible && (
        <LinkIssueUI
          searchJiraLink={searchJiraLink}
          onOk={issueUrl => {
            onLinkIssue(issueUrl);
            setLinkIssueVisible(false);
          }}
          onCancel={() => setLinkIssueVisible(false)}
        />
      )}

      {issues.length > 0 && (
        <div
          style={{
            width: '100%',
            display: 'flex',
            marginBottom: '10px',
            marginTop: '10px',
          }}
        >
          <Dropdown
            overlay={
              <Menu>
                <Menu.Item
                  key="create"
                  onClick={() => setCreateIssueVisible(true)}
                >
                  Create issue
                </Menu.Item>
                <Menu.Item key="link" onClick={() => setLinkIssueVisible(true)}>
                  Link issue
                </Menu.Item>
              </Menu>
            }
          >
            <Button type="primary" style={{ marginLeft: 'auto' }}>
              Add <DownOutlined />
            </Button>
          </Dropdown>
        </div>
      )}

      {issues.length === 0 && (
        <Empty description="No linked issues.">
          <Button
            type="primary"
            onClick={() => setCreateIssueVisible(true)}
            role="button"
            name="Create Issue"
          >
            Create Issue
          </Button>{' '}
          or{' '}
          <Button
            type="default"
            onClick={() => setLinkIssueVisible(true)}
            role="button"
            name="Link Issue"
          >
            Link Existing Issue
          </Button>
        </Empty>
      )}

      {issues.length > 0 && (
        <Table
          size="small"
          dataSource={issues}
          columns={[
            displayType === 'project'
              ? {
                  title: 'Link type',
                  render: issue => (issue.resourceId ? 'Resource' : 'Project'),
                }
              : {},
            {
              title: 'Issue',
              render: issue => {
                return (
                  <a href={issue.url} target="_blank">
                    {issue.summary}
                  </a>
                );
              },
            },
            displayType === 'project'
              ? {
                  title: 'Resource',
                  render: issue =>
                    issue.resourceSelfUrl && (
                      <a
                        onClick={() =>
                          onNavigateToResource &&
                          onNavigateToResource(issue.resourceSelfUrl)
                        }
                      >
                        {issue.resourceLabel}
                      </a>
                    ),
                }
              : {},
            {
              title: 'Last updated',
              render: issue => {
                return <>{getFriendlyTimeAgoString(issue.updated)}</>;
              },
            },
            {
              title: 'Status',
              render: issue => {
                return <>{issue.status}</>;
              },
            },
            {
              title: 'Comments',
              render: issue => {
                return <>{issue.commentCount}</>;
              },
            },
            {
              key: 'id',
              render: issue => {
                return (
                  <>
                    <Dropdown.Button
                      size="small"
                      overlay={
                        <Menu>
                          <Menu.Item
                            key="unlink"
                            onClick={() => confirmUnlinkIssue(issue.key)}
                          >
                            Unlink
                          </Menu.Item>
                        </Menu>
                      }
                    >
                      Options
                    </Dropdown.Button>
                  </>
                );
              },
            },
          ]}
        />
      )}
    </>
  );
};

export default JIRAPluginUI;
export { AuthorizeJiraUI };
