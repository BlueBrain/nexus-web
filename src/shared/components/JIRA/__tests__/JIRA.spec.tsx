import * as React from 'react';
import { NexusProvider } from '@bbp/react-nexus';
import { createNexusClient } from '@bbp/nexus-sdk';
import JIRA from '../JIRA';
import { QueryClient, QueryClientProvider } from 'react-query';
import fetch from 'node-fetch';
import {
  render,
  fireEvent,
  waitFor,
  screen,
  server,
} from '../../../../utils/testUtil';
import { rest } from 'msw';
import '@testing-library/jest-dom';
import { act } from 'react-dom/test-utils';
describe('JIRA', () => {
  it('renders with empty prop values', () => {
    const { container } = render(
      <JIRA
        projects={[]}
        issues={[]}
        onCreateIssue={jest.fn()}
        onLinkIssue={jest.fn()}
        onUnlinkIssue={jest.fn()}
        searchJiraLink={''}
        displayType="resource"
        onNavigateToResource={jest.fn()}
        isLoading={false}
      ></JIRA>
    );
    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('No linked issues.');
    expect(container).toHaveTextContent('Create Issue');
    expect(container).toHaveTextContent('Link Existing Issue');
  });
  it('renders with issues', () => {
    const issues = [
      {
        resourceId: 'testId',
        summary: 'An issue summary',
        resourceLabel: 'self url',
        updated: '2020-01-08',
        status: '',
        commentCount: 0,
        key: 'blah',
      },
    ];
    const { container } = render(
      <JIRA
        projects={[]}
        issues={issues}
        onCreateIssue={jest.fn()}
        onLinkIssue={jest.fn()}
        onUnlinkIssue={jest.fn()}
        searchJiraLink={''}
        displayType="resource"
        onNavigateToResource={jest.fn()}
        isLoading={false}
      ></JIRA>
    );
    expect(container).toMatchSnapshot();
    expect(container).not.toHaveTextContent('No linked issues.');
    expect(container).not.toHaveTextContent('Create Issue');
    expect(container).not.toHaveTextContent('Link Existing Issue');
  });
  it('renders with projects', () => {
    const issues = [
      {
        summary: 'An issue summary',
        resourceLabel: 'self url',
        updated: '2020-01-08',
        status: '',
        commentCount: 0,
        key: 'blah',
      },
    ];
    const { container } = render(
      <JIRA
        projects={[]}
        issues={issues}
        onCreateIssue={jest.fn()}
        onLinkIssue={jest.fn()}
        onUnlinkIssue={jest.fn()}
        searchJiraLink={''}
        displayType="project"
        onNavigateToResource={jest.fn()}
        isLoading={false}
      ></JIRA>
    );
    expect(container).toMatchSnapshot();
    expect(container).not.toHaveTextContent('No linked issues.');
    expect(container).not.toHaveTextContent('Create Issue');
    expect(container).not.toHaveTextContent('Link Existing Issue');
  });
  xit('renders with projects', async () => {
    const createIssue = jest.fn();
    const { container } = render(
      <JIRA
        projects={[]}
        issues={[]}
        onCreateIssue={createIssue}
        onLinkIssue={jest.fn()}
        onUnlinkIssue={jest.fn()}
        searchJiraLink={''}
        displayType="resource"
        onNavigateToResource={jest.fn()}
        isLoading={false}
      ></JIRA>
    );
    const createButton = await screen.findByText('Create Issue');
    act(() => {
      fireEvent.click(createButton);
    });
    screen.debug();
    const createIssueModal = screen.getAllByAltText(
      'A Jira issue will be created and linked to this Nexus'
    );
    expect(createIssueModal).toHaveLength(1);
  });
});
