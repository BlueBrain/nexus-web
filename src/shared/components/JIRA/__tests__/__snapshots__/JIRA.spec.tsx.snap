// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JIRA > renders authraizeJIRAUI 1`] = `
<div>
  <h2>
    We first need to connect your Jira account.
  </h2>
  <ol>
    <li>
      Visit Jira to log-in and authorize access to Jira from Nexus Fusion
      <br />
      <br />
      <a
        href="jirurri"
        target="_blank"
      >
        Authorize Jira access
      </a>
      <br />
      <br />
    </li>
    <li>
      Copy the 
      <em>
        verification code
      </em>
       from step 1 into the text box below and click Connect.
      <br />
      <br />
      <div
        style="display: flex;"
      >
        <div>
          <input
            class="ant-input css-dev-only-do-not-override-17a39f8"
            placeholder="Verification code"
            style="background-color: rgb(255, 255, 255); width: 200px;"
            type="text"
            value=""
          />
        </div>
        <div>
          <button
            class="ant-btn css-dev-only-do-not-override-17a39f8 ant-btn-primary"
            type="button"
          >
            <span>
              Connect
            </span>
          </button>
        </div>
      </div>
    </li>
  </ol>
</div>
`;

exports[`JIRA > renders with empty prop values 1`] = `
<div>
  <div
    class="css-dev-only-do-not-override-17a39f8 ant-empty"
  >
    <div
      class="ant-empty-image"
    >
      <svg
        height="152"
        viewBox="0 0 184 152"
        width="184"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <g
            transform="translate(24 31.67)"
          >
            <ellipse
              cx="67.797"
              cy="106.89"
              fill="#F5F5F7"
              fill-opacity=".8"
              rx="67.797"
              ry="12.668"
            />
            <path
              d="M122.034 69.674L98.109 40.229c-1.148-1.386-2.826-2.225-4.593-2.225h-51.44c-1.766 0-3.444.839-4.592 2.225L13.56 69.674v15.383h108.475V69.674z"
              fill="#AEB8C2"
            />
            <path
              d="M101.537 86.214L80.63 61.102c-1.001-1.207-2.507-1.867-4.048-1.867H31.724c-1.54 0-3.047.66-4.048 1.867L6.769 86.214v13.792h94.768V86.214z"
              fill="url(#linearGradient-1)"
              transform="translate(13.56)"
            />
            <path
              d="M33.83 0h67.933a4 4 0 0 1 4 4v93.344a4 4 0 0 1-4 4H33.83a4 4 0 0 1-4-4V4a4 4 0 0 1 4-4z"
              fill="#F5F5F7"
            />
            <path
              d="M42.678 9.953h50.237a2 2 0 0 1 2 2V36.91a2 2 0 0 1-2 2H42.678a2 2 0 0 1-2-2V11.953a2 2 0 0 1 2-2zM42.94 49.767h49.713a2.262 2.262 0 1 1 0 4.524H42.94a2.262 2.262 0 0 1 0-4.524zM42.94 61.53h49.713a2.262 2.262 0 1 1 0 4.525H42.94a2.262 2.262 0 0 1 0-4.525zM121.813 105.032c-.775 3.071-3.497 5.36-6.735 5.36H20.515c-3.238 0-5.96-2.29-6.734-5.36a7.309 7.309 0 0 1-.222-1.79V69.675h26.318c2.907 0 5.25 2.448 5.25 5.42v.04c0 2.971 2.37 5.37 5.277 5.37h34.785c2.907 0 5.277-2.421 5.277-5.393V75.1c0-2.972 2.343-5.426 5.25-5.426h26.318v33.569c0 .617-.077 1.216-.221 1.789z"
              fill="#DCE0E6"
            />
          </g>
          <path
            d="M149.121 33.292l-6.83 2.65a1 1 0 0 1-1.317-1.23l1.937-6.207c-2.589-2.944-4.109-6.534-4.109-10.408C138.802 8.102 148.92 0 161.402 0 173.881 0 184 8.102 184 18.097c0 9.995-10.118 18.097-22.599 18.097-4.528 0-8.744-1.066-12.28-2.902z"
            fill="#DCE0E6"
          />
          <g
            fill="#FFF"
            transform="translate(149.65 15.383)"
          >
            <ellipse
              cx="20.654"
              cy="3.167"
              rx="2.849"
              ry="2.815"
            />
            <path
              d="M5.698 5.63H0L2.898.704zM9.259.704h4.985V5.63H9.259z"
            />
          </g>
        </g>
      </svg>
    </div>
    <div
      class="ant-empty-description"
    >
      No linked issues.
    </div>
    <div
      class="ant-empty-footer"
    >
      <button
        class="ant-btn css-dev-only-do-not-override-17a39f8 ant-btn-primary"
        name="Create Issue"
        role="button"
        type="button"
      >
        <span>
          Create Issue
        </span>
      </button>
       
      or
       
      <button
        class="ant-btn css-dev-only-do-not-override-17a39f8 ant-btn-default"
        name="Link Issue"
        role="button"
        type="button"
      >
        <span>
          Link Existing Issue
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`JIRA > renders with issues 1`] = `
<div>
  <div
    style="width: 100%; display: flex; margin-bottom: 10px; margin-top: 10px;"
  >
    <button
      class="ant-btn css-dev-only-do-not-override-17a39f8 ant-btn-primary ant-dropdown-trigger"
      style="margin-left: auto;"
      type="button"
    >
      <span>
        Add 
      </span>
      <span
        aria-label="down"
        class="anticon anticon-down"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="down"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    class="ant-table-wrapper css-dev-only-do-not-override-17a39f8"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-17a39f8"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table ant-table-small"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup />
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <td
                      class="ant-table-cell"
                    />
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Issue
                    </th>
                    <td
                      class="ant-table-cell"
                    />
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Last updated
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Status
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Comments
                    </th>
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="blah"
                  >
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    >
                      <a
                        target="_blank"
                      >
                        An issue summary
                      </a>
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    >
                      4 years ago
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      <div
                        class="ant-space-compact css-dev-only-do-not-override-17a39f8 ant-space-compact-block ant-dropdown-button css-dev-only-do-not-override-17a39f8"
                      >
                        <button
                          class="ant-btn css-dev-only-do-not-override-17a39f8 ant-btn-default ant-btn-sm ant-btn-compact-item ant-btn-compact-first-item"
                          type="button"
                        >
                          <span>
                            Options
                          </span>
                        </button>
                        <button
                          class="ant-btn css-dev-only-do-not-override-17a39f8 ant-btn-default ant-btn-sm ant-btn-icon-only ant-btn-compact-item ant-btn-compact-last-item ant-dropdown-trigger"
                          type="button"
                        >
                          <span
                            class="ant-btn-icon"
                          >
                            <span
                              aria-label="ellipsis"
                              class="anticon anticon-ellipsis"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="ellipsis"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
                                />
                              </svg>
                            </span>
                          </span>
                        </button>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <ul
          class="ant-pagination ant-pagination-mini ant-table-pagination ant-table-pagination-right css-dev-only-do-not-override-17a39f8"
        >
          <li
            aria-disabled="true"
            class="ant-pagination-prev ant-pagination-disabled"
            title="Previous Page"
          >
            <button
              class="ant-pagination-item-link"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="left"
                class="anticon anticon-left"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                  />
                </svg>
              </span>
            </button>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
            tabindex="0"
            title="1"
          >
            <a
              rel="nofollow"
            >
              1
            </a>
          </li>
          <li
            aria-disabled="true"
            class="ant-pagination-next ant-pagination-disabled"
            title="Next Page"
          >
            <button
              class="ant-pagination-item-link"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="right"
                class="anticon anticon-right"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`JIRA > renders with projects 1`] = `
<div>
  <div
    style="width: 100%; display: flex; margin-bottom: 10px; margin-top: 10px;"
  >
    <button
      class="ant-btn css-dev-only-do-not-override-17a39f8 ant-btn-primary ant-dropdown-trigger"
      style="margin-left: auto;"
      type="button"
    >
      <span>
        Add 
      </span>
      <span
        aria-label="down"
        class="anticon anticon-down"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="down"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    class="ant-table-wrapper css-dev-only-do-not-override-17a39f8"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-17a39f8"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table ant-table-small"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup />
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Link type
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Issue
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Resource
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Last updated
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Status
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Comments
                    </th>
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="blah"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      Project
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      <a
                        target="_blank"
                      >
                        An issue summary
                      </a>
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    >
                      4 years ago
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      <div
                        class="ant-space-compact css-dev-only-do-not-override-17a39f8 ant-space-compact-block ant-dropdown-button css-dev-only-do-not-override-17a39f8"
                      >
                        <button
                          class="ant-btn css-dev-only-do-not-override-17a39f8 ant-btn-default ant-btn-sm ant-btn-compact-item ant-btn-compact-first-item"
                          type="button"
                        >
                          <span>
                            Options
                          </span>
                        </button>
                        <button
                          class="ant-btn css-dev-only-do-not-override-17a39f8 ant-btn-default ant-btn-sm ant-btn-icon-only ant-btn-compact-item ant-btn-compact-last-item ant-dropdown-trigger"
                          type="button"
                        >
                          <span
                            class="ant-btn-icon"
                          >
                            <span
                              aria-label="ellipsis"
                              class="anticon anticon-ellipsis"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="ellipsis"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
                                />
                              </svg>
                            </span>
                          </span>
                        </button>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <ul
          class="ant-pagination ant-pagination-mini ant-table-pagination ant-table-pagination-right css-dev-only-do-not-override-17a39f8"
        >
          <li
            aria-disabled="true"
            class="ant-pagination-prev ant-pagination-disabled"
            title="Previous Page"
          >
            <button
              class="ant-pagination-item-link"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="left"
                class="anticon anticon-left"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                  />
                </svg>
              </span>
            </button>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
            tabindex="0"
            title="1"
          >
            <a
              rel="nofollow"
            >
              1
            </a>
          </li>
          <li
            aria-disabled="true"
            class="ant-pagination-next ant-pagination-disabled"
            title="Next Page"
          >
            <button
              class="ant-pagination-item-link"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="right"
                class="anticon anticon-right"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`JIRA > shows loading spinner 1`] = `
<div>
  <div
    style="width: 100%; display: flex; margin-bottom: 10px; margin-top: 10px; height: 200px; justify-content: center; align-items: center;"
  >
    <div
      aria-busy="true"
      aria-live="polite"
      class="ant-spin ant-spin-lg ant-spin-spinning ant-spin-show-text css-dev-only-do-not-override-17a39f8"
    >
      <span
        class="ant-spin-dot ant-spin-dot-spin"
      >
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
      </span>
    </div>
  </div>
</div>
`;
